src/adapters/bidderFactory.js

```diff
@@ -0,0 +1,240 @@
+      const placementCodesHandled = {};
+      function addBidWithCode(placementCode, bid) {
+        placementCodesHandled[placementCode] = true;
+        bidmanager.addBidResponse(placementCode, bid);
+      }
+      function fillNoBids() {
+        bidsRequest.bids
+          .map(bidRequest => bidRequest.placementCode)
+          .forEach(placementCode => {
+            if (placementCode && !placementCodesHandled[placementCode]) {
+              bidmanager.addBidResponse(placementCode, newEmptyBid());
+            }
+          });
+      }
+
+      const bidRequests = bidsRequest.bids.filter(filterAndWarn);
+      if (bidRequests.length === 0) {
+        fillNoBids();
+        return;
+      }
+
+      let requests = spec.buildRequests(bidRequests);
+      if (!requests || requests.length === 0) {
+        fillNoBids();
+        return;
+      }
+      if (!Array.isArray(requests)) {
+        requests = [requests];
+      }
+
+      // Callbacks don't compose as nicely as Promises. We should call fillNoBids() once _all_ the
+      // Server requests have returned and been processed. Since `ajax` accepts a single callback,
+      // we need to rig up a function which only executes after all the requests have been responded.
+      const onResponse = delayExecution(fillNoBids, requests.length)
+      requests.forEach(processRequest);
+
+      function processRequest(request) {
+        switch (request.type) {
+          case GET:
+            ajax(
+              `${request.endpoint}?${parseQueryStringParameters(request.data)}`,
+              {
+                success: onSuccess,
+                error: onFailure
+              },
+              undefined,
+              {
+                method: GET,
+                withCredentials: true
+              }
+            );
+            break;
+          case POST:
+            ajax(
+              request.endpoint,
+              {
+                success: onSuccess,
+                error: onFailure
+              },
+              typeof request.data === 'string' ? request.data : JSON.stringify(request.data),
+              {
+                method: POST,
+                contentType: 'text/plain',
+                withCredentials: true
+              }
+            );
+            break;
+          default:
+            logWarn(`Skipping invalid request from ${spec.code}. Request type ${request.type} must be GET or POST`);
+            onResponse();
+        }
+      }
+
+      // If the server responds successfully, use the adapter code to unpack the Bids from it.
+      // If the adapter code fails, no bids should be added. After all the bids have been added, make
+      // sure to call the `onResponse` function so that we're one step closer to calling fillNoBids().
+      function onSuccess(response) {
+        const bidRequestMap = {};
+        bidRequests.forEach(bid => {
+          bidRequestMap[bid.bidId] = bid;
+        });
```

Comment:
Why is this `bidRequestMap` recreated for each successful request?  Couldn't it just be made once in the same context as the `bidRequests`?